src/hotspot/share/compiler/compileBroker.cpp

```diff
@@ -443,6 +443,11 @@ CompileTask* CompileQueue::get(CompilerThread* thread) {
     AbstractCompiler* compiler = thread->compiler();
     guarantee(compiler != nullptr, "Compiler object must exist");
     compiler->on_empty_queue(this, thread);
+    if (_first != nullptr) {
+      // The call to on_empty_queue may have temporarily unlocked the MCQ lock
+      // so check again whether any tasks were added to the queue.
+      break;
+    }
```

Comment:
So this is just optimization to avoid waiting 5 sec?

Reply:
Yes. Normally it does not matter but in `-Xbatch` mode it means one compilation every 5 seconds and tests like `TestTrichotomyExpressions` time out.